libp2p/protocols/secure/secure.nim

```diff
@@ -44,24 +46,23 @@ proc readLoop(sconn: SecureConn, stream: BufferStream) {.async.} =
   except CatchableError as exc:
     trace "Exception occurred Secure.readLoop", exc = exc.msg
   finally:
+    trace "closing conn", closed = conn.closed()
+    if not conn.closed:
+      await conn.close()
+
+    trace "closing sconn", closed = sconn.closed()
     if not sconn.closed:
       await sconn.close()
-    trace "ending Secure readLoop", isclosed = sconn.closed()
+    trace "ending Secure readLoop"
 
 proc handleConn*(s: Secure, conn: Connection, initiator: bool = false): Future[Connection] {.async, gcsafe.} =
   var sconn = await s.handshake(conn, initiator)
   proc writeHandler(data: seq[byte]) {.async, gcsafe.} =
     trace "sending encrypted bytes", bytes = data.shortLog
     await sconn.writeMessage(data)
 
-  var stream = newBufferStream(writeHandler)
-  asyncCheck readLoop(sconn, stream)
-  result = newConnection(stream)
-  result.closeEvent.wait()
-    .addCallback do (udata: pointer):
-      trace "wrapped connection closed, closing upstream"
-      if not isNil(sconn) and not sconn.closed:
-        asyncCheck sconn.close()
+  result = newConnection(newBufferStream(writeHandler))
+  asyncCheck readLoop(sconn, result)
```

Comment:
I kinda wanna remove this too and track the future but I guess can be done in another PR

Reply:
this needs to happen in a more focused refactor, we'll do that after we get the current implementation more stable